src/Compilers/Core/Portable/Syntax/GreenNode.cs

```diff
@@ -1017,5 +1017,61 @@ internal GreenNode AddError(DiagnosticInfo err)
             // Get a new green node with the errors added on.
             return SetDiagnostics(errorInfos);
         }
+
+        /// <summary>
+        /// Walks the green tree from this node downwards and lazily returns all
+        /// diagnostics found.
+        /// </summary>
+        internal virtual IEnumerable<DiagnosticInfo> GetAllSyntaxErrors()
+        {
+            if (ContainsDiagnostics)
+            {
+                foreach (var diagnostic in GetDiagnostics())
+                {
+                    yield return diagnostic;
+                }
+
+                for (int i = 0; i < SlotCount; i++)
+                {
+                    GreenNode child = GetSlot(i);
+                    foreach (var childDiagnostic in child.GetAllSyntaxErrors())
```

Comment:
>child.GetAllSyntaxErrors() [](start = 52, length = 26)

It feels like we might end up allocating a lot of enumerators for a sub-tree with many levels of nesting. It looks like VB implementation was intentionally avoiding that.